library-best-practices/JLBP-4.md

```diff
@@ -0,0 +1,57 @@
+[JLBP-4] Avoid dependencies on unstable libraries and features
+--------------------------------------------------------------
+
+- At a high level, don't depend on things that can have frequent breaking
+  changes (where frequent means more than once every couple years).
+- The primary type of unstable library is one that allows for breaking changes
+  to their Public API within the same major version. For libraries following
+  semver, this means libraries with a 0.x.y version. (See [JLBP-3](JLBP-3.md)
+  for more details on the recommendations for following semver.)
+- Another type of unstable library is one that provides stable versions
+  according to semver, but which bumps major versions frequently.
+  - The primary example of this is Guava before version 22.0, which published a
+    new major version once or more per year.
+- Unstable features are features that are not part of the stable Public API of a
+  stable library, often marked with annotations like `@Beta`.  (See
+  [JLBP-3](JLBP-3.md) for more details on the recommendations for annotating
+  unstable features.)
+- If your library depends on an unstable library or feature, and that feature
+  experiences a breaking change between versions, your library will be locked to
+  a specific version of your dependency.
+  - If you expose the unstable feature on your library's surface, then your
+    library's current major version will be permanently locked to the version
+    you initially exposed, and you won't be able to upgrade without making a
+    breaking change to your users.
+  - If you only use the unstable feature in your implementation, then each minor
+    or patch version of your library will require a very specific version of
+    your dependency, and it will be unsafe for your users to upgrade your
+    library on its own, creating opportunities for hard-to-diagnose runtime
+    conflicts for users.
```

Comment:
```
  - If your library's release cadence does not match that of your dependencies,
    then your library may be the cause of dependency conflicts for consumers of your library.
```

This is the problem we've seen with Eclipse projects, where we were shipping 3 different versions of Guava at one point.

Reply:
What exactly do you mean by release cadence here? I don't think having a dependency with a faster release cadence of minor version bumps would be a problem - it seems that this only matters for major version bumps.

Reply:
Release cadence = frequency of making releases.  Each release introduces risk of breaking change — sometimes inadvertent, but sometimes deliberate — to un- or under-documented behaviour.